src/lib/statistics/statistics_objects/range_filter.cpp

```diff
@@ -270,43 +254,55 @@ bool RangeFilter<T>::does_not_contain(const PredicateCondition predicate_conditi
     case PredicateCondition::BetweenLowerExclusive:
     case PredicateCondition::BetweenUpperExclusive:
     case PredicateCondition::BetweenExclusive: {
-      /* There are two scenarios where a between predicate can be pruned:
-       *    - both bounds are "outside" (not spanning) the segment's value range (i.e., either both are smaller than
-       *      the minimum or both are larger than the maximum
-       *    - both bounds are within the same gap
-       */
+      // There are two scenarios where a between predicate can be pruned:
+      //    (i) Both bounds are "outside" (not spanning) the segment's value range (i.e., either both are smaller than
+      //        the minimum or both are larger than the maximum).
+      //   (ii) Both bounds are within the same gap.
 
-      Assert(variant_value2, "Between operator needs two values.");
+      Assert(variant_value2, "Between predicate needs two values.");
       const auto value2 = boost::get<T>(*variant_value2);
 
-      // a BETWEEN 5 AND 4 will always be empty
+      // a BETWEEN 5 AND 4 will always be empty.
       if (value2 < value) {
         return true;
       }
 
+      const auto is_lower_inclusive = is_lower_inclusive_between(predicate_condition);
+      const auto is_upper_inclusive = is_upper_inclusive_between(predicate_condition);
+
       // Smaller than the segment's minimum.
-      if (does_not_contain(PredicateCondition::LessThanEquals, std::max(value, value2))) {
+      const auto min = ranges.front().first;
+      const auto is_below_min = is_upper_inclusive_between(predicate_condition) ? (value2 < min) : (value2 <= min);
+      if (is_below_min) {
         return true;
       }
 
       // Larger than the segment's maximum.
-      if (does_not_contain(PredicateCondition::GreaterThanEquals, std::min(value, value2))) {
+      const auto max = ranges.back().second;
+      const auto is_above_max = is_lower_inclusive_between(predicate_condition) ? (value > max) : (value >= max);
+      if (is_above_max) {
         return true;
       }
 
-      const auto range_comp = [](const std::pair<T, T> range, const T compare_value) {
-        return range.second < compare_value;
-      };
-      // Get value range or next larger value range if searched value is in a gap.
-      const auto start_lower = std::lower_bound(ranges.cbegin(), ranges.cend(), value, range_comp);
-      const auto end_lower = std::lower_bound(ranges.cbegin(), ranges.cend(), value2, range_comp);
+      // Find the range that starts at an element >= value2.
+      const auto upper_bound_range =
+          std::lower_bound(ranges.cbegin(), ranges.cend(), value2,
+                           [](const auto& range, const auto compare_value) { return range.first < compare_value; });
+
+      // If we find such a range, check for inclusive case.
+      auto upper_bound_in_gap =
+          upper_bound_range == ranges.cend() || !is_upper_inclusive || upper_bound_range->first > value2;
```

Comment:
Why does `is_upper_inclusive` being false lead to `upper_bound_in_gap` being true?

Reply:
I'm gonna add some more comments (maybe a running example), but the idea is the following: With `lower_bound`, we find the first range $r_i$ s.t. the range min >= value2. If the predicate is not upper inclusive, value2 is always in the gap brtween $r_{i-1}$ and $r_i$. Otherwise (predicate is inclusive), we need to check if value2 > $\min r_i$ to see if it is in the gap.

Reply:
Is the explanation with the example sufficient?

Reply:
Yes, sounds good.